pkg/auth/requests/token_review.go

```diff
@@ -0,0 +1,78 @@
+package requests
+
+import (
+	"net/http"
+
+	"github.com/rancher/rancher/pkg/auth/tokens"
+	"github.com/rancher/steve/pkg/auth"
+	"github.com/sirupsen/logrus"
+	v1 "k8s.io/api/authentication/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apiserver/pkg/authentication/user"
+	authv1 "k8s.io/client-go/kubernetes/typed/authentication/v1"
+)
+
+var (
+	allowedPaths = map[string]bool{
+		// This is to enable prometheus to use
+		// mounted service account tokens to
+		// fetch metrics.
+		"/metrics": true,
+	}
+)
+
+type TokenReviewAuth struct {
+	AuthClient           authv1.AuthenticationV1Interface
+	InitialAuthenticator auth.Authenticator
+}
+
+func NewTokenReviewAuth(authenticator auth.Authenticator, authClient authv1.AuthenticationV1Interface) auth.Authenticator {
+	return &TokenReviewAuth{
+		AuthClient:           authClient,
+		InitialAuthenticator: authenticator,
+	}
+}
+
+// Authenticate attempts to authenticate using given function. If authentication
+// fails AND the endpoint is in allowedPaths, it will attempt to perform a token
+// review to authenticate token and extract user info.
+func (t *TokenReviewAuth) Authenticate(req *http.Request) (user.Info, bool, error) {
+	info, hasAuth, err := t.InitialAuthenticator.Authenticate(req)
+	if err != nil {
+		return nil, false, err
+	}
+
+	if info.GetName() != "system:cattle:error" {
```

Comment:
Also, I don't think this logic works since we don't parse a user to `system:cattle:error` during the Authenticate call; it happens after Authenticate is called in the Middleware function. Therefore, I think the code as it stands right now would never hit this point since it would have failed out with the original error.

I think this needs to be modified to:
```suggestion
	if hasAuth {
```

Reply:
I think this happens because of the tokenAuthenticator that all requests go through. This code has been tested and when the token is invalid this is the name.

Reply:
Nothing gets set as a result of the initial Authenticate because it is not wrapped in the ToMiddleware func which is where that happens. So the info set is from a previous handler. I am changing this in my refactor to be more straight forward.